strutil/shlex/shlex_test.go

```diff
@@ -157,3 +157,69 @@ func TestNastyReader(t *testing.T) {
 		t.Errorf("unexpected error")
 	}
 }
+
+func TestQuote(t *testing.T) {
+	safeUnquoted := "abcdefghijklmnopqrstuvwxyz" + "ABCDEFGHIJKLMNOPQRSTUVWXYZ" + "0123456789" + "@%_-+=:,./"
+	unicodeSample := "\xe9\xe0\xdf" // e + acute accent, a + grave, sharp s
+	unsafe := "\"`$\\!" + unicodeSample
+
+	type testInfo struct {
+		input    string
+		expected string
+	}
+
+	quoteTests := []testInfo{{
+		input:    ``,
+		expected: `''`,
+	}, {
+		input:    safeUnquoted,
+		expected: safeUnquoted,
+	}, {
+		input:    `foo bar xyz`,
+		expected: `'foo bar xyz'`,
+	}}
+
+	for _, u := range unsafe {
+		quoteTests = append(quoteTests, testInfo{
+			input:    `foo` + string(u) + `bar`,
+			expected: `'foo` + string(u) + `bar'`,
+		})
+		quoteTests = append(quoteTests, testInfo{
+			input:    `foo` + string(u) + `'bar'`,
+			expected: `'foo` + string(u) + `'"'"'bar'"'"''`,
+		})
+	}
+
+	for _, test := range quoteTests {
+		quoted := Quote(test.input)
+		if quoted != test.expected {
+			t.Errorf("expected %s, got %s", test.expected, quoted)
+		}
+	}
+}
+
+func TestJoin(t *testing.T) {
+	joinTests := []struct {
+		input    []string
+		expected string
+	}{{
+		input:    []string{`a `, `b`},
+		expected: `'a ' b`,
+	}, {
+		input:    []string{`a`, ` b`},
+		expected: `a ' b'`,
+	}, {
+		input:    []string{`a`, ` `, `b`},
+		expected: `a ' ' b`,
+	}, {
+		input:    []string{`"a`, `b"`},
+		expected: `'"a' 'b"'`,
+	}}
```

Comment:
I think we should also test the edge cases of empty slice and nil slice (should be same as empty). Probably also a case where a single argument is like `x y`, and where a single is more than one character like `/foo/bar`. We don't have to be exhaustive as other things should be caught in the `Quote` tests, but just a couple more cases here would be good.

Reply:
Added a few cases more. Thank you.